tests/unit/development/test_collect_nodes.py

```diff
@@ -0,0 +1,62 @@
+from dash.development._collect_nodes import collect_nodes
+
+metadata = {
+    "string": {"type": {"name": "string"}},
+    "shape": {
+        "type": {
+            "name": "shape",
+            "value": {
+                "single": {"type": {"name": "number"}},
+                "node": {"type": {"name": "node"}},
+            },
+        }
+    },
+    "list_of_nodes": {
+        "type": {"name": "arrayOf", "value": {"name": "node"}},
+    },
+    "list_of_union": {
+        "type": {
+            "name": "arrayOf",
+            "value": {
+                "name": "union",
+                "value": [
+                    {
+                        "name": "shape",
+                        "value": {
+                            "a": {"type": {"name": "string"}},
+                            "b": {"type": {"name": "element"}},
+                        },
+                    },
+                    {"name": "node"},
+                ],
+            },
+        }
+    },
+    "list_of_shapes": {
+        "type": {
+            "name": "arrayOf",
+            "value": {
+                "name": "shape",
+                "value": {"label": {"name": "node"}, "value": {"name": "string"}},
+            },
+        }
+    },
+    "mixed": {
+        "type": {"name": "union", "value": [{"name": "number"}, {"name": "element"}]}
+    },
+    "direct": {"type": {"name": "node"}},
+}
+
+
+def test_dcn001_collect_nodes():
+    nodes = collect_nodes(metadata)
+
+    assert nodes == [
+        "shape.node",
+        "list_of_nodes",
+        "[]list_of_union.b",
+        "[]list_of_union",
+        "[]list_of_shapes.label",
```

Comment:
Putting the `[]` at the front may be limiting eventually, it obscures the difference between `a.b[0].c` and `a[0].b.c`

```py
>>> metadata = {"a": {"type": {"name": "shape", "value": {"b": {"type": {"name": "arrayOf", "value": {"name": "shape", "value": {"c": {"name": "node"}}}}}}}}}
>>> collect_nodes(metadata)[0]
'[]a.b.c'
>>> md2 = {"a": {"type": {"name": "arrayOf", "value": {"name": "shape", "value": {"b": {"type": {"name": "shape", "value": {"c": {"name": "node"}}}}}}}}}
>>> collect_nodes(md2)[0]
'[]a.b.c'
```
whereas if the brackets go after the attribute, these would look like `a.b[].c` and `a[].b.c`. In principle you could also imagine `a[].b[].c` vs `a[][].b.c` and so on.

Maybe not an issue in the short term, but I feel like with just that change of notation we could reach complete generality of anywhere a component might allow nodes.

Reply:
Changed the notation of the brackets to the back.